### PR TITLE
Add: rust-lsp

### PR DESCRIPTION
**Pull request by jamartin9**
Add rust-lsp completion with company, flycheck and bindings.
Edit: Updated with typo fix
refs: #10443 #11161

### CODE REVIEW

**Comment by jvillasante** [comment]
Is this one already merged to develop? I've been waiting for it!

**Comment by matthewjberger** [comment]
This PR works wonderfully

**Comment by sdwolfz** [comment]
Can you please give a short description of why you chose racer as a backend by default? The PR looks good otherwise (except those closing parenthesis that need to be on the same line as everything else).

**Comment by jamartin9** [comment]
@sdwolfz I chose racer as the default backend to not radically change existing behavior for the users of the layer who may not want to have this customization option forced on them. For example the use of this backend requires the lsp layer and rustup/rls. Current users of the rust layer may not have these installed and this would most likely break their current setup. 

tl;dr To minimize potentially breaking changes. 

Would you like this description in the commit message, comment or just here on the PR?

I think I fixed most of the parentheses. Feel free to point out any i missed and I'll address them too.

File: layers/+lang/rust/funcs.el [hunk 1]
```diff
@@ -24,10 +24,77 @@ open a junk Rust file, type in some code and quickly run it.
...
+                       buffer)
+                     (message "Compilation failed with: %s" status))))
     (compile
-     (format "rustc -o %s %s && %s"
-             (shell-quote-argument output-file-name)
-             (shell-quote-argument input-file-name)
-             (shell-quote-argument output-file-name)))))
+      (format "rustc -o %s %s"
+              (shell-quote-argument output-file-name)
+              (shell-quote-argument input-file-name)))))
```
[the remark this answers is not in the document]
Comment (jamartin9) [reply layers/+lang/rust/funcs.el]: It changes the quick run function to use compile hooks to run vs using && ( which didn't work for me on some shell I was testing ). I'll remove it and just keep it on a diff branch for me to use until #9535 is resolved

**Comment by jamartin9** [comment]
@sdwolfz Unfortunately I do not have a simple project that would demonstrate this well. However it seems to work on most libraries/projects I've tried. The changes in this layer are mostly to take advantage of the `lsp` layer handling the configuration of the ui, jump bindings, flycheck and company. The closest to a simple setup I can come up with for this off the top of my head is roughly
```
# install rustup
curl https://sh.rustup.rs -sSf | sh
rustup default nightly
rustup update
# install rls
rustup component add rls-preview rust-analysis rust-src
# create project
cargo new --bin hello_world
# open source
emacs hello_world/src/main.rs
```
Having the following in your `dotspacemacs-configuration-layers` :
```
auto-completion
lsp
(rust :variables
    rust-backend 'lsp)
```
rls will put the files under `hello_world/target/rls`
It might take a few moments on projects with lots of dependencies for the initial run.
Seems to work with remote dependencies built and installed with cargo but I'll limit this example to the standard library. Change hello_world/src/main.rs to:
```
use std::env;
fn main() {
let var = "s";
env::set_var(var, var);
assert_eq!(env::var(var), Ok(var.to_string()));
} 
```
By default I think you can use `Alt .` and `Alt ,` to jump to/back on definitions. I turn company on with `(global-company-mode t)` other options are available from the `lsp` layer.

**Comment by sdwolfz** [comment]
Thank you :heart:! 
And congratulations on your first Spacemacs PR :tada:!

I was not able to test it out locally as `rustup component add rls-preview rust-analysis rust-src` gave me `error: component 'rls-preview' for 'x86_64-unknown-linux-gnu' is unavailable for download`. But judging from the positive feedback this PR received, and from looking through the code I see no reason to keep it waiting. If there is a problem I'll be quick to respond.

Cherry-picked into develop branch, you can safely delete your branch.

**Comment by yetone** [comment]
@sdwolfz This is a great job! Because of nightly updated frequently, so RLS component always gone away. Yesterday RLS in nightly is break, but today it is recover.

https://rust-lang-nursery.github.io/rust-toolstate/

**Comment by XiangQingW** [comment]
@jamartin9 This is an amazing PR, but I stuck in some trouble. When I press M-.  sometime emacs prompts: No definitions found for xxx. Please do me a favor, thank you very much.

**Comment by jamartin9** [comment]
@XiangQingW Without a specific example to test I can only give general advice.
Some of the standard library jump-to-definitions don't seem to work properly at times but this is usually an upstream issue with the way `rls` provides the data or the way `rust-lsp` retrieves/displays it. 

The actual code in the PR was just adding `rust-lsp` to spacemaces rust layer. Bugs in that or its depenencies are still going to be present.

1. Make sure `rls` `racer` and your emacs packages are up to date. (RLS still uses `racer` according to its README.md so ensure that is configured properly aswell.)
2. Check for relevant issues in the upstream repos ( ie are you using workspaces or some other specific feature?) 
3. Ensure `rls` is actually running (it can crash too).
4. Try to do a `cargo build` of the project to ensure the dependencies are downloaded and wait a few moments for rls to index them (can get no definitions if rls hasn't indexed the lib yet). 

If you want to link me to a specific example to replicate I can see if it is reproducible and chase it upstream if needed.

**Comment by XiangQingW** [comment]
@jamartin9 Thanks for replying. I'll try those advices.
